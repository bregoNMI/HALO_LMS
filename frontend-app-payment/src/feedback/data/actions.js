export const ADD_MESSAGE = 'ADD_MESSAGE';
export const REMOVE_MESSAGE = 'REMOVE_MESSAGE';
export const CLEAR_MESSAGES = 'CLEAR_MESSAGES';

export const addMessage = (code, userMessage, data, messageType, fieldName = null) => ({
  type: ADD_MESSAGE,
  payload: {
    code,
    userMessage,
    data,
    messageType,
    fieldName,
  },
});

export const removeMessage = id => ({
  type: REMOVE_MESSAGE,
  payload: {
    id,
  },
});

export const clearMessages = () => ({
  type: CLEAR_MESSAGES,
});
