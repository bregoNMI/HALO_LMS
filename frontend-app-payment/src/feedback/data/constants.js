export const ALERT_TYPES = {
  DANGER: 'danger',
  SUCCESS: 'success',
  INFO: 'info',
  WARNING: 'warning',
};

export const MESSAGE_TYPES = {
  DEBUG: 'debug',
  INFO: 'info',
  SUCCESS: 'success',
  WARNING: 'warning',
  ERROR: 'error',
};
