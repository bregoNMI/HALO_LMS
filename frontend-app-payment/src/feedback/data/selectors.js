import { createSelector } from 'reselect';

import { storeName } from './reducers';

const messagesSelector = state => state[storeName];

export const messageListSelector = createSelector(
  messagesSelector,
  messages => messages.orderedIds.map(id => messages.byId[id]),
);

export const alertListMapStateToProps = createSelector(
  messageListSelector,
  messageList => ({
    messageList,
  }),
);

const messagesAsArraySelector = createSelector(
  messagesSelector,
  messages => Object.values(messages.byId),
);

const fieldNameSelector = (state, props) => props.name;

/**
 * In order to use this selector, a component must have a string "name" prop that tells the selector
 * which field to return the messages for.
 */
export const fieldMessagesSelector = createSelector(
  fieldNameSelector,
  messagesAsArraySelector,
  (fieldName, messages) => {
    const fieldMessages = [];
    messages.forEach((message) => {
      if (message.fieldName === fieldName) {
        fieldMessages.push(message);
      }
    });
    return fieldMessages;
  },
);
