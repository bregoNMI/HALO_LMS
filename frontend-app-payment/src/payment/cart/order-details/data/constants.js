// Product types
export const COURSE_ENTITLEMENT_PRODUCT_TYPE = 'Course Entitlement';
export const ENROLLMENT_CODE_PRODUCT_TYPE = 'Enrollment Code';
export const SEAT_PRODUCT_TYPE = 'Seat';

// Certificate types
export const VERIFIED_CERTIFICATE_TYPE = 'verified';
export const CREDIT_CERTIFICATE_TYPE = 'credit';

// Order Details message keys
export const ENTITLEMENT_MESSAGE_KEY = 'entitlement';
export const ENROLLMENT_CODE_MESSAGE_KEY = 'enrollment.code';
export const VERIFIED_SEAT_MESSAGE_KEY = 'seat.verified';
export const CREDIT_SEAT_MESSAGE_KEY = 'seat.credit';
export const SEAT_MESSAGE_KEY = 'seat';
