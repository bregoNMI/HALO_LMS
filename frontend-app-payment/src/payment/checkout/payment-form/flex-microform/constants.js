export const STATUS_LOADING = 'LOADING';
export const STATUS_READY = 'READY';
export const DEFAULT_STATUS = STATUS_LOADING;
