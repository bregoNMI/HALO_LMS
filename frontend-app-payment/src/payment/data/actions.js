import { createRoutine } from 'redux-saga-routines';

// Routines are action + action creator pairs in a series.
// Actions adhere to the flux standard action format.
// Routines by default are in the form of:
//
// Action                |   Action Creator
// -----------------------------------------------
// fetchBasket.TRIGGER   |   fetchBasket()
// fetchBasket.SUCCESS   |   fetchBasket.success()
// fetchBasket.FAILURE   |   fetchBasket.failure()
// fetchBasket.FULFILL   |   fetchBasket.fulfill()
//
// Created with redux-saga-routines
export const fetchCaptureKey = createRoutine('FETCH_CAPTURE_KEY');
export const fetchClientSecret = createRoutine('FETCH_CLIENT_SECRET');
export const submitPayment = createRoutine('SUBMIT_PAYMENT');
export const fetchBasket = createRoutine('FETCH_BASKET');
export const addCoupon = createRoutine('ADD_COUPON');
export const removeCoupon = createRoutine('REMOVE_COUPON');
export const updateQuantity = createRoutine('UPDATE_QUANTITY');

// Actions and their action creators
export const BASKET_DATA_RECEIVED = 'BASKET_DATA_RECEIVED';

export const basketDataReceived = basket => ({
  type: BASKET_DATA_RECEIVED,
  payload: basket,
});

export const BASKET_PROCESSING = 'BASKET_PROCESSING';

export const basketProcessing = isProcessing => ({
  type: BASKET_PROCESSING,
  payload: isProcessing,
});

export const CAPTURE_KEY_PROCESSING = 'CAPTURE_KEY_PROCESSING';

export const captureKeyProcessing = isProcessing => ({
  type: CAPTURE_KEY_PROCESSING,
  payload: isProcessing,
});

export const CLIENT_SECRET_PROCESSING = 'CLIENT_SECRET_PROCESSING';

export const clientSecretProcessing = isProcessing => ({
  type: CLIENT_SECRET_PROCESSING,
  payload: isProcessing,
});

export const MICROFORM_STATUS = 'MICROFORM_STATUS';

export const microformStatus = status => ({
  type: MICROFORM_STATUS,
  payload: status,
});

export const CAPTURE_KEY_START_TIMEOUT = 'CAPTURE_KEY_START_TIMEOUT';

export const captureKeyStartTimeout = () => ({
  type: CAPTURE_KEY_START_TIMEOUT,
});
export const CAPTURE_KEY_DATA_RECEIVED = 'CAPTURE_KEY_DATA_RECEIVED';

export const captureKeyDataReceived = captureKey => ({
  type: CAPTURE_KEY_DATA_RECEIVED,
  payload: captureKey,
});

export const CLIENT_SECRET_DATA_RECEIVED = 'CLIENT_SECRET_DATA_RECEIVED';

export const clientSecretDataReceived = clientSecret => ({
  type: CLIENT_SECRET_DATA_RECEIVED,
  payload: clientSecret,
});
