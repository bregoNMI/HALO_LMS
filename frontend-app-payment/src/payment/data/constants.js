export const ORDER_TYPES = {
  BULK_ENROLLMENT: 'Enrollment Code',
  ENTITLEMENT: 'Course Entitlement',
  SEAT: 'Seat',
};

export const CERTIFICATE_TYPES = {
  VERIFIED: 'verified',
  CREDIT: 'credit',
};
