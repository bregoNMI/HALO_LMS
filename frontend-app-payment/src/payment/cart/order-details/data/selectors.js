import { createSelector, createStructuredSelector } from 'reselect';

import { getModuleState } from '../../../data/utils';
import {
  SEAT_MESSAGE_KEY,
  CREDIT_SEAT_MESSAGE_KEY,
  VERIFIED_SEAT_MESSAGE_KEY,
  SEAT_PRODUCT_TYPE,
  VERIFIED_CERTIFICATE_TYPE,
  CREDIT_CERTIFICATE_TYPE,
  ENROLLMENT_CODE_PRODUCT_TYPE,
  COURSE_ENTITLEMENT_PRODUCT_TYPE,
  ENTITLEMENT_MESSAGE_KEY,
  ENROLLMENT_CODE_MESSAGE_KEY,
} from './constants';

export const storePath = ['payment', 'basket'];

const paymentSelector = (state, props) => getModuleState(state, props.storePath ? props.storePath : storePath);

const lastProductSelector = createSelector(
  paymentSelector,
  ({ products }) => products[products.length - 1],
);

const messageTypeSelector = createSelector(
  lastProductSelector,
  (lastProduct) => {
    if (lastProduct === undefined) {
      return null;
    }
    const { productType, certificateType } = lastProduct;
    if (productType === COURSE_ENTITLEMENT_PRODUCT_TYPE) {
      return ENTITLEMENT_MESSAGE_KEY;
    }
    if (productType === ENROLLMENT_CODE_PRODUCT_TYPE) {
      return ENROLLMENT_CODE_MESSAGE_KEY;
    }
    if (productType === SEAT_PRODUCT_TYPE) {
      if (certificateType === VERIFIED_CERTIFICATE_TYPE) {
        return VERIFIED_SEAT_MESSAGE_KEY;
      }
      if (certificateType === CREDIT_CERTIFICATE_TYPE) {
        return CREDIT_SEAT_MESSAGE_KEY;
      }
      return SEAT_MESSAGE_KEY;
    }
    return null;
  },
);

export const orderDetailsMapStateToProps = createStructuredSelector({
  messageType: messageTypeSelector,
});
