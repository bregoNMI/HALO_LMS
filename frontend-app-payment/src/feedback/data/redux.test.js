import { createStore, combineReducers } from 'redux';

import reducer, { resetIds } from './reducers';
import { addMessage, removeMessage, clearMessages } from './actions';
import { MESSAGE_TYPES } from './constants';
import { messageListSelector, alertListMapStateToProps, fieldMessagesSelector } from './selectors';

describe('redux tests', () => {
  let store;
  beforeEach(() => {
    store = createStore(combineReducers({ feedback: reducer }));
    resetIds();
  });

  it('should return the default state when appropriate', () => {
    // Its base state
    expect(store.getState().feedback).toMatchSnapshot();

    // If an action isn't appropriate for it
    store.dispatch({ type: 'SOMETHING_ELSE' });
    expect(store.getState().feedback).toMatchSnapshot();

    // When called with no parameters
    expect(reducer()).toMatchSnapshot();
  });

  it('should select the right data', () => {
    // Selectors
    let result = messageListSelector(store.getState());
    expect(result).toEqual([]);

    result = alertListMapStateToProps(store.getState());
    expect(result).toEqual({ messageList: [] });

    result = fieldMessagesSelector(store.getState(), { name: 'boo' });
    expect(result).toEqual([]);
  });

  describe('addMessage action', () => {
    it('should add a code-based message', () => {
      store.dispatch(addMessage('boo', null, { needed: 'data' }, MESSAGE_TYPES.ERROR));
      expect(store.getState().feedback.byId).toEqual({
        0: {
          id: 0,
          code: 'boo',
          userMessage: null,
          data: { needed: 'data' },
          messageType: MESSAGE_TYPES.ERROR,
          fieldName: null,
        },
      });
      expect(store.getState().feedback.orderedIds).toEqual([0]);

      store.dispatch(addMessage('boo_again', null, { needed: 'data1' }, MESSAGE_TYPES.INFO));
      expect(store.getState().feedback.byId).toEqual({
        0: {
          id: 0,
          code: 'boo',
          userMessage: null,
          data: { needed: 'data' },
          messageType: MESSAGE_TYPES.ERROR,
          fieldName: null,
        },
        1: {
          id: 1,
          code: 'boo_again',
          userMessage: null,
          data: { needed: 'data1' },
          messageType: MESSAGE_TYPES.INFO,
          fieldName: null,
        },
      });
      expect(store.getState().feedback.orderedIds).toEqual([0, 1]);
    });

    it('should add user_message-based message', () => {
      store.dispatch(addMessage(null, 'oh no oh no', {}, MESSAGE_TYPES.ERROR));
      expect(store.getState().feedback.byId).toEqual({
        0: {
          id: 0,
          code: null,
          userMessage: 'oh no oh no',
          data: {},
          messageType: MESSAGE_TYPES.ERROR,
          fieldName: null,
        },
      });
    });

    it('should add field-based message', () => {
      store.dispatch(addMessage(null, 'oh no oh no', {}, MESSAGE_TYPES.ERROR, 'cats'));
      expect(store.getState().feedback.byId).toEqual({
        0: {
          id: 0,
          code: null,
          userMessage: 'oh no oh no',
          data: {},
          messageType: MESSAGE_TYPES.ERROR,
          fieldName: 'cats',
        },
      });
    });

    it('should select the messages', () => {
      store.dispatch(addMessage('boo', null, { needed: 'data' }, MESSAGE_TYPES.ERROR));
      store.dispatch(addMessage('boo2', null, { needed: 'data2' }, MESSAGE_TYPES.ERROR, 'bah'));

      const expectedMessageList = [
        {
          id: 0,
          code: 'boo',
          userMessage: null,
          data: { needed: 'data' },
          messageType: MESSAGE_TYPES.ERROR,
          fieldName: null,
        },
        {
          id: 1,
          code: 'boo2',
          userMessage: null,
          data: { needed: 'data2' },
          messageType: MESSAGE_TYPES.ERROR,
          fieldName: 'bah',
        },
      ];

      let result = messageListSelector(store.getState());
      expect(result).toEqual(expectedMessageList);

      result = alertListMapStateToProps(store.getState());
      expect(result).toEqual({ messageList: expectedMessageList });

      result = fieldMessagesSelector(store.getState(), { name: 'bah' });
      expect(result).toEqual([{
        id: 1,
        code: 'boo2',
        userMessage: null,
        data: { needed: 'data2' },
        messageType: MESSAGE_TYPES.ERROR,
        fieldName: 'bah',
      }]);
    });
  });

  describe('removeMessage action', () => {
    beforeEach(() => {
      store.dispatch(addMessage('boo0', null, null, MESSAGE_TYPES.INFO));
    });

    it('should remove the only message present', () => {
      store.dispatch(removeMessage(0));
      expect(store.getState().feedback).toEqual({
        byId: {},
        orderedIds: [],
      });
    });

    it('should remove the first message', () => {
      store.dispatch(addMessage('boo1', null, null, MESSAGE_TYPES.INFO));
      store.dispatch(addMessage('boo2', null, null, MESSAGE_TYPES.INFO));
      store.dispatch(removeMessage(0));
      expect(store.getState().feedback).toEqual({
        byId: {
          1: {
            id: 1,
            code: 'boo1',
            userMessage: null,
            data: null,
            messageType: MESSAGE_TYPES.INFO,
            fieldName: null,
          },
          2: {
            id: 2,
            code: 'boo2',
            userMessage: null,
            data: null,
            messageType: MESSAGE_TYPES.INFO,
            fieldName: null,
          },
        },
        orderedIds: [1, 2],
      });
    });

    it('should remove the middle message', () => {
      store.dispatch(addMessage('boo1', null, null, MESSAGE_TYPES.INFO));
      store.dispatch(addMessage('boo2', null, null, MESSAGE_TYPES.INFO));
      store.dispatch(removeMessage(1));
      expect(store.getState().feedback).toEqual({
        byId: {
          0: {
            id: 0,
            code: 'boo0',
            userMessage: null,
            data: null,
            messageType: MESSAGE_TYPES.INFO,
            fieldName: null,
          },
          2: {
            id: 2,
            code: 'boo2',
            userMessage: null,
            data: null,
            messageType: MESSAGE_TYPES.INFO,
            fieldName: null,
          },
        },
        orderedIds: [0, 2],
      });
    });

    it('should remove the last message', () => {
      store.dispatch(addMessage('boo1', null, null, MESSAGE_TYPES.INFO));
      store.dispatch(addMessage('boo2', null, null, MESSAGE_TYPES.INFO));
      store.dispatch(removeMessage(2));
      expect(store.getState().feedback).toEqual({
        byId: {
          0: {
            id: 0,
            code: 'boo0',
            userMessage: null,
            data: null,
            messageType: MESSAGE_TYPES.INFO,
            fieldName: null,
          },
          1: {
            id: 1,
            code: 'boo1',
            userMessage: null,
            data: null,
            messageType: MESSAGE_TYPES.INFO,
            fieldName: null,
          },
        },
        orderedIds: [0, 1],
      });
    });
  });

  describe('clearMessages action', () => {
    beforeEach(() => {
      store.dispatch(addMessage('boo0', null, null, MESSAGE_TYPES.INFO));
      store.dispatch(addMessage('boo1', null, null, MESSAGE_TYPES.ERROR));
    });

    it('should clear all messages', () => {
      store.dispatch(clearMessages());
      expect(store.getState().feedback).toEqual({
        byId: {},
        orderedIds: [],
      });
    });
  });
});
