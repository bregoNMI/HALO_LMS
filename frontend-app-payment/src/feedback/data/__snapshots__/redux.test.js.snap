// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux tests should return the default state when appropriate 1`] = `
Object {
  "byId": Object {},
  "orderedIds": Array [],
}
`;

exports[`redux tests should return the default state when appropriate 2`] = `
Object {
  "byId": Object {},
  "orderedIds": Array [],
}
`;

exports[`redux tests should return the default state when appropriate 3`] = `
Object {
  "byId": Object {},
  "orderedIds": Array [],
}
`;
